### PR TITLE
Better handling of options for unfolding/theoryAgnostic analyses

### PR DESCRIPTION
**Pull request by cippy**
**NOTE** (added on 16/03/2024): because of technical complications I decided to give up with the subparsers, so the text below is partially obsolete. The current strategy is based on a new option --analysisMode (with default to None for the traditional analysis) and accepting arguments from the previously devised subparsers name (see below). Essentially I just update the main parser adding the relevant options, defined in a function specifically for this purpose. This function is called when defining the main parser, so no need to call it explicitly.

This PR improves upon previously merged PR#402
Subparsers are implement to better organise the several options we have, ensuring that only those options relevant for a specific analysis can be called. Also, options that are in common for unfolding and theory agnostic are defined in a common place and propagated to the specific subparser.
In principle this PR should not touch the physics, only the way to deal with the code.

Notable differences wrt before are as follows:
- options --unfolding/--theoryAgnostic/--theoryAgnosticPolVar are removed and now have to be called with the subparser keywords (unfolding, theoryAgnosticNormVar, theoryAgnosticPolVar)
-  the previous --theoryAgnostic option is now mapped into theoryAgnosticNormVar, so to better categorise wrt to other versions (currently the only alternative is theoryAgnosticPolVar, or the original version with floating POIs which is done when not using option --poiAsNoi)
- option --genVars in the histmakers was renamed --genAxes to be consistent with the equivalent option in setupCombine 

Other features to note:
- the subparser gets defined after the main parser, and can thus be invoked anywhere in the command string, not necessarily at the end. This is because the main parser is passed to the subparser as a parent (so the subparser knows all options of the main parser, as long as they are not defined after the subparser itself).
- if an option with nargs='+' is called right before the subparser is invoked in the command string, the subparser name will be counted as one of nargs elements. Apparently there is no way to solve it with parsers in python. Keep that in mind if the subparser is not invoked in the beginning of the command string.

Some further cleanup is still possible. In particular, the current theoryAgnostic versions assume that one always wants to use option --poiAsNoi (it is not true by default because unfolding can use both cases), otherwise the original version with floating POIs is obtained (which was never fully implemented for all systematic uncertainties, and presumably will no longer be used in any case).

### CODE REVIEW

**Comment by cippy** [comment]
I guess the main reviewers for this PR should be @davidwalter2 and @emanca 

I remind you that the main modification is that the unfolding and theory agnostic analyses should now get called with --analysisMode [unfolding | theoryAgnosticNormVar] --poiAsNoi (there is also the third option theoryAgnosticPolVar).

In fact, after dropping the sub parsers, I might restore the original options --unfolding and --theoryAgnosticNormVar, but given the similarities I thought that it could be cleaner to have --analysisMode option so we don't need to define exceptions or checks against calling them together

File: scripts/combine/setupCombine.py [hunk 1]
```diff
@@ -832,36 +861,44 @@ def outputFolderName(outfolder, card_tool, doStatOnly, postfix):
...
+    cardTool.writeOutput(args=args, forceNonzero=forceNonzero, check_systs=checkSysts)
     return
 
 if __name__ == "__main__":
     parser = make_parser()
     args = parser.parse_args()
     
     logger = logging.setup_logger(__file__, args.verbose, args.noColorLogger)
+    
+    isUnfolding = args.analysisMode == "unfolding"
```
Comment (davidwalter2) [inline scripts/combine/setupCombine.py:879]: Is this block duplicated? it also appears in L126-132

Comment (cippy) [reply scripts/combine/setupCombine.py:879]: yeah it is, unfortunately there is no way to propagate all those flags in a simple way, passing them to the function as arguments is also not a good idea

Comment (cippy) [reply scripts/combine/setupCombine.py:879]: I could define them as global variables to set only once, if you prefer, but usually global variables are deprecated

Comment (davidwalter2) [reply scripts/combine/setupCombine.py:879]: No it's ok, fine from my side to merge

**Pull request merged**